students/tri_nguyen/lesson04/mailroom.py

```diff
@@ -0,0 +1,175 @@
+
+        if full_name.lower() == 'list':
+            display_donor_list()
+            continue
+
+        elif full_name not in name_list:
+            donation_amount = float(input('Enter donation amount: '))
+            donor_dict.append(dict(name=full_name.title(), donation=donation_amount, count=1, average=donation_amount))
+
+        elif full_name in name_list:
+            donation_amount = float(input('Enter donation amount: '))
+            for item in donor_dict:
+                if item['name'] == full_name:
+                    item['donation'] = item['donation'] + donation_amount
+                    item['count'] = item['count'] + 1
+                    item['average'] = item['donation'] / item['count']
+
+        choice = input('Do you want to quit and send email?(y/n) ')
+
+        if choice == 'y':
+            email_template = '''
+                Dear {0},
+
+                Thank you for giving us ${1:,.2f}. Your money will be put to good use.
+                Please come back and donate more.
+
+                Best Regards,
+
+                John Doe
+            '''
+            print(email_template.format(full_name, float(donation_amount)))
+
+            break
+        else:
+            continue
+
+
+def create_report():
+    ''' this function generate a report of donors '''
+
+    template = '{0:<26}|{1:^13}|{2:^11}|{3:^14}'
+    value_template = '{0:<26} ${1:12.2f} {2:11} ${3:13.2f}'
+
+    formatted_columns = template.format('Donor Name', 'Total Given', 'Num Gifts', 'Average Gift')
+
+    print(formatted_columns)
+
+    print('{}'.format('-' * len(formatted_columns)))
+
+    print('Report of Donors is below:\n')
+
+    donor_list = [list(item.values()) for item in donor_dict]
+
+    donor_list.sort(key=itemgetter(1), reverse=True)
+
+    for item in donor_list:
+        print(value_template.format(item[0], item[1], item[2], item[3]))
+    print()
+
+
+def send_all_letters():
+    ''' this function send letters to everyone '''
+
+    email_dict = {
+        'greetings': 'Dear',
+        'body_one': 'Thank you for your kind donation of ',
+        'body_two': 'It will be put to very good use.',
+        'end': 'Sincerely',
+        'signature': '-The Team'}
+
+    value_list = [list(item.values()) for item in donor_dict]
+
+    for each_list in value_list:
+
+        each_list[0] = each_list[0].split()
+
+        template = '{}_' * (len(each_list[0]) - 1) + '{}.txt'
+
+        first_name = each_list[0][0]
+        last_name = each_list[0][1]
+
+        with open(template.format(*each_list[0]), 'w') as f:
+            f.write('{0} {1} {2},\n\n'.format(email_dict['greetings'], first_name, last_name))
+            f.write('\t{0}${1:,.2f}\n\n'.format(email_dict['body_one'], each_list[1]))
+            f.write('\t{}\n\n'.format(email_dict['body_two']))
+            f.write('\t\t{}\n'.format(email_dict['end']))
+            f.write('\t\t    {}'.format(email_dict['signature']))
```

Comment:
I wouldn't use `email_dict` as a way to populate the file content string. 

what makes a bit more sense is to have a template at the top of the file
```
first_name = "natasha"
template = "Dear {first_name}, thank you for your donation!"
# then in the loop:
template.format(first_name=first_name)
```
the way it is right now makes it kind of hard to read

Reply:
Having the template at the top of  the file meaning outside of the function  at the top of the script?

Reply:
sorry I meant around line 119 outside the loop but inside of `send_all_letters` function